Added CustomQABot and CustomQABotAllFeatures. (#3712)
microsoft/BotBuilder-Samples: samples/csharp_dotnetcore/48.customQABot-all-features/README.md

# Custom Question Answering

> Bot Framework v4 Custom question answering bot sample. This sample shows how to integrate Multiturn and Active learning in a Custom Question Answering bot with ASP.Net Core-2. Click [here](https://docs.microsoft.com/en-us/azure/cognitive-services/language-service/question-answering/tutorials/guided-conversations) to know more about using follow-up prompts to create multiturn conversation. To know more about how to enable and use active learning, click [here](https://docs.microsoft.com/en-us/azure/cognitive-services/language-service/question-answering/tutorials/active-learning).

Reviewer: Rephrase for CQA

## Concepts introduced in this sample
The [Custom question answering feature in Language Service][LS] enables you to build, train and publish a simple question and answer bot based on FAQ URLs, structured documents or editorial content in minutes.
In this sample, we demonstrate 
- How to use the Active Learning to generate suggestions for knowledge base.
- How to use follow up prompts to create multiple turns of a conversation.
- How to configure display of precise answers.
- How to enable/disable querying unstructured sources with the bot.

# Prerequisites
- Create a [Language resource](https://aka.ms/create-language-resource) with Custom question answering enabled.
- Follow instructions [here][Quickstart] to create a Custom question answering project. You will need this project's name to be used as `QnAKnowledgebaseId` in [appsettings.json](appsettings.json).

### Update [appsettings.json](appsettings.json) to connect your bot to the knowledge base
- In the [Azure Portal][Azure], go to your resource.
- Go to Keys and Endpoint under Resource Management.

> - `QnAEndpointKey` would be one of the keys and `QnAEndpointHostName` would be the Endpoint from [Azure Portal][Azure].

Reviewer: these lines, please change to as suggested in CustomQABot readme.

    ```

- In a terminal, navigate to `samples/csharp_dotnetcore/48.customQABot-all-features`
- Run the bot from a terminal or from Visual Studio, choose option A or B.

  A) From a terminal

  ```bash
  # run the bot
  dotnet run
  ```

  B) Or from Visual Studio

  - Launch Visual Studio
  - File -> Open -> Project/Solution
  - Navigate to `samples/csharp_dotnetcore/48.customQABot-all-features` folder
  - Select `CustomQABotAllFeatures.csproj` file
  - Press `F5` to run the project
- Connect to the bot using Bot Framework Emulator
  1) Launch Bot Framework Emulator
  2) File -> Open Bot
  3) Enter a Bot URL of `http://localhost:3978/api/messages`

# Try Active Learning
- Try the following queries:
  1) Surface
  2) Features
- In Language Studio, click on inspect to see closeness in scores of returned answers.
- In [Bot Framework Emulator][BFE], a card is generated with suggestions shown.
  - Clicking an option would send a [feedback record](https://docs.microsoft.com/en-us/rest/api/cognitiveservices/questionanswering/question-answering-projects/add-feedback) which would show as suggestion under `Review suggestions` in [Language Studio][LS].

>   - `DefaultCardTitle`, `DefaultCardNoMatchText` and `DefaultCardNoMatchResponse` in the card could be changed from [RootDialog.cs](Dialogs/RootDialog.cs).

Reviewer: names changed now?

- Sample queries:
  1) Frontline workers
  2) Hybrid work solutions
- You can observe that, answers are returned with high score.
- You can set `IncludeUnstructuredSources` to false in [RootDialog.cs](Dialogs/RootDialog.cs) to prevent querying unstructured sources.

# Try Filters
- Go to your project in [Language Studio][LS] -> In `Edit knowledge bases` -> Under **Metadata** column click on `+ Add`
- Add a key value pair and click on `Save changes`.
- Click on `Test` and add metadata that you just added by clicking on **Show advanced options**.
- This will query only those qnas with the given metadata.
- You can also pass filters in [rootDialog.cs](dialogs/rootDialog.cs)(Line No. 79) and apply source filters and logical operations. [Learn more](https://docs.microsoft.com/en-us/rest/api/cognitiveservices/questionanswering/question-answering/get-answers#queryfilters).

# Microsoft Teams channel group chat fix
When a bot (named as `HelpBot`) is added to a Teams channel or Teams group chat, you will have to refer it as `@HelpBot question to bot` to get answers from the service.

> However, bot tries to send `<at>HelpBot</at> question to bot` as query to Custom question answering service which may not give expected results for question to bot. The following code removes <at>HelpBot</at> mentions of the bot from the message and sends the remaining text as query to the service.

Reviewer: `question to bot` in ticks

Reviewer: `<at>HelpBot</at>` in back ticks

> # Custom Question Answering

Author: I updated readme of 12.customQABot with the suggestions below.

- Go to `Edit knowledge base` -> Click on `...` -> Click on `Import questions and answers` -> Click on `Import as TSV`.
- Import [SampleForCQA.tsv](CognitiveModels/SampleForCQA.tsv) file.
- You can test your bot by clicking on `Test` option.
- Go to `Deploy knowledge base` and click on `Deploy`.

# To try this sample

- Install the Bot Framework Emulator version 4.14.0 or greater from [here][BFE]
- Clone the repository

    ```bash
    git clone https://github.com/Microsoft/botbuilder-samples.git
    ```

- In a terminal, navigate to `samples/csharp_dotnetcore/48.customQABot-all-features`
- Run the bot from a terminal or from Visual Studio, choose option A or B.

  A) From a terminal

  ```bash
  # run the bot
  dotnet run
  ```

  B) Or from Visual Studio

  - Launch Visual Studio
  - File -> Open -> Project/Solution
  - Navigate to `samples/csharp_dotnetcore/48.customQABot-all-features` folder
  - Select `CustomQABotAllFeatures.csproj` file
  - Press `F5` to run the project
- Connect to the bot using Bot Framework Emulator
  1) Launch Bot Framework Emulator
  2) File -> Open Bot
  3) Enter a Bot URL of `http://localhost:3978/api/messages`

# Try Active Learning

> - Try the following queries:

Reviewer 2: ```suggestion
- Try the following utterances:
```

  1) Surface
  2) Features

> - In Language Studio, click on inspect to see closeness in scores of returned answers.

Reviewer 2: ```suggestion
- In Language Studio, click on inspect to see the closeness in the scores of the returned answers.
```

> - In [Bot Framework Emulator][BFE], a card is generated with suggestions shown.

Reviewer 2: ```suggestion
- In [Bot Framework Emulator][BFE], a card is generated with the suggestions.
```

  - Clicking an option would send a [feedback record](https://docs.microsoft.com/en-us/rest/api/cognitiveservices/questionanswering/question-answering-projects/add-feedback) which would show as suggestion under `Review suggestions` in [Language Studio][LS].
  - `ActiveLearningCardTitle`, `ActiveLearningCardNoMatchText` and `ActiveLearningCardNoMatchResponse` in the card could be changed from [RootDialog.cs](Dialogs/RootDialog.cs).

# Try Multi-turn prompt

> - Try the following queries:

Reviewer 2: ```suggestion
- Try the following utterances:
```

  1) Accessibility
  2) Options

> - You will notice that multi-turn prompts associated with the question are also returned with response.

Reviewer 2: ```suggestion
- You will notice that multi-turn prompts associated with the question are also returned in the responses.
```

# Try Precise Answering
- Try the following queries:
  1) Accessibility
  2) Features
- You can notice a short answer returned along with a long answer.
- If testing in [Language Studio][LS], you might have to check `Include short answer response` at the top.
- You can disable precise answering by setting `EnablePreciseAnswer` to false in [appsettings.json](appsettings.json).
- You can set `DisplayPreciseAnswerOnly` in [appsettings.json](appsettings.json) to true to display just precise answers in the response.

# Query unstructured content
- Go to your project in [Language Studio][LS] -> In `Manage sources` click on `+ Add source`

> - Click on `URLs` and add [this](https://www.microsoft.com/en-us/microsoft-365/blog/2022/01/27/from-empowering-frontline-workers-to-accessibility-improvements-heres-whats-new-in-microsoft-365/) link by classifying it as **unstructured** under `Classify file structure`.

Reviewer 2: ```suggestion
- Click on `URLs` and add `https://www.microsoft.com/en-us/microsoft-365/blog/2022/01/27/from-empowering-frontline-workers-to-accessibility-improvements-heres-whats-new-in-microsoft-365/1` and select **unstructured** in the `Classify file structure` dropdown.
```
